Connect props

Mikeysax opened the pull request:
Connected redux store to App.js props so they can be passed down to Children. Added Posts, NotFound components, and posts reducer with corresponding tests. There seems to be a bug within:

`React.cloneElement` 

Getting:

 `TypeError: Cannot read property 'props' of undefined`

Looking for answers on Stack...

Commit:
Changed spec to test for consistency and..

added NotFound component with testing. NotFound component was imported into App.js and is used for other routes which are not specified.

Commit:
Setup store and temporary data...

for testing the connection of props between children. Created container to bind store data to AppContainer props and pass to ApplicationLayout for children. Created index action to aggrigate all future actions.

Commit:
Added posts reducer and Post component

Commit:
Fixed react bug with ApplicationLayout component saying ref and key error, changed Welcome to Home, added more tests for Posts component

Commit:
bug in testing React.cloneElement, asking on SO for possible solution

Commit:
Added more tests, removed action index test because it contains no logic.

Mikeysax commented:
@lasley @jonmast @dino217 Thoughts?

Commit:
Reverted to previous component routing and layout with bug present in testing

Commit:
Changed store connection

Removed app container and replaced connection to Home. Refactor of tests to reflect changes with Home as a container. ApplicationLayout acts as a simple component now which displays children.

Commit:
Working Home Container test added

Commit:
Added MIT License and README description

Commit:
Fixed node version in package.json for codeship CI

Commit:
Fixed node version in package.json for codeship CI

Commit:
Added markdown badge and fixed package.json after noticing the ability to change the node version under codeship CI test settings

Mikeysax commented:
@lasley @jonmast @dino217 @dakotabryant @rjdellecese @kenmazaika How's this look?

kenmazaika commented on src/App.js:
```diff
@@ -18,7 +19,8 @@ const router = (
       <Route path="/" component={ApplicationLayout}>
-        <IndexRoute component={Welcome} />
+        <IndexRoute component={Home} />
+        <Route path="*" component={NotFound} />
```
So a bit of a nit-pick here, but if your react app responds to that, I think we'd need some server logic to properly respond with a 404 not found in this case.  Setthing that up is way beyond the scope of what I understand about server/client rendering of JSX.

Mikeysax replied:
https://github.com/reactjs/react-router/blob/b60d6c0351ff91cf04bccdac8c4b6e976aec94ec/upgrade-guides/v1.0.0.md#notfound-route

 Thoughts?

kenmazaika commented on src/js/containers/Home.js:
```diff
@@ -0,0 +1,34 @@
+    return (
+      <div>
+        <div className="col-xs-10 col-xs-offset-1 col-sm-8 col-sm-offset-2 col-md-6 col-md-offset-3">
+          {this.props.posts.map((post, i) => <Post key={i} i={i} post={post} />)}
```
IMO I would move this out to outside the JSX embedded into a `const`.  I've seen this style though and think it's popular too.  I also don't like the prop of `i`.  `index` or `postIndex` might be more clear.

kenmazaika commented on src/store.js:
```diff
@@ -7,14 +7,23 @@ import thunk from 'redux-thunk';
+    "comment": "Singing is amazing.",
+    "id":"2"
+  }
+]
```
Semicolon missing.

![image](https://cloud.githubusercontent.com/assets/233615/15691980/ce2aded8-2759-11e6-909f-ae7355346843.png)

kenmazaika replied:
Is this the same format as the response from the API will be?  We should make sure there is parity, as it will make integration MUCH easier.

kenmazaika commented on src/store.js:
```diff
@@ -7,14 +7,23 @@ import thunk from 'redux-thunk';
+const posts = [
+  {
+    "name": "John Doe",
+    "comment": "I wish I was a real person.",
```
solid.

kenmazaika commented on tests/components/NotFound.test.js:
```diff
@@ -0,0 +1,21 @@
+import NotFound from '../../src/js/components/NotFound';
+
+describe('NotFound Component', () => {
+  var notFoundComponent
```
![image](https://cloud.githubusercontent.com/assets/233615/15692027/144f5a24-275a-11e6-9d53-f192021e4164.png)

kenmazaika commented on tests/components/Post.test.js:
```diff
@@ -0,0 +1,26 @@
+import Post from '../../src/js/components/Post';
+
+describe('Post Component', () => {
+  var postComponent
```
![image](https://cloud.githubusercontent.com/assets/233615/15692033/1cd96a04-275a-11e6-8643-c741b703c491.png)

kenmazaika commented on src/App.js:
```diff
@@ -7,7 +7,8 @@ import './css/application.scss';
 import ApplicationLayout from './js/components/ApplicationLayout';
-import Welcome from './js/components/Welcome';
+import Home from './js/containers/Home';
+import NotFound from './js/components/NotFound'
```
Inconsistent semicolons on import statements.

![image](https://cloud.githubusercontent.com/assets/233615/15692050/395d78be-275a-11e6-922e-013162cd5962.png)

kenmazaika commented on tests/components/ApplicationLayout.test.js:
```diff
@@ -0,0 +1,16 @@
+import ApplicationLayout from '../../src/js/components/ApplicationLayout';
+
+describe('ApplicationLayout Component', () => {
+  var appLayoutComponent
```
![image](https://cloud.githubusercontent.com/assets/233615/15692068/589df370-275a-11e6-96ea-9434be12607f.png)

kenmazaika commented on tests/components/Navbar.test.js:
```diff
@@ -0,0 +1,25 @@
+import Navbar from '../../src/js/components/Navbar';
+
+describe('Navbar Component', () => {
+  var navbarComponent
```
![image](https://cloud.githubusercontent.com/assets/233615/15692070/5d9be09e-275a-11e6-9ec6-ef16dcd9389b.png)

kenmazaika commented on tests/containers/Home.test.js:
```diff
@@ -0,0 +1,23 @@
+import Home from '../../src/js/containers/Home';
+
+describe('Home Container', () => {
+  var homeContainer
```
![image](https://cloud.githubusercontent.com/assets/233615/15692075/65024008-275a-11e6-9a25-914bba24c298.png)

kenmazaika commented:
Few nitpicks from me, but I think it looks solid.  Looks like a really good first pass, and solid job setting some testing up there too.

Mikeysax commented:
Thanks Ken! I love that picture. Derppin' with the semi-colons.

kenmazaika commented:
That's Douglas Crockford.  He made jslint and I'm pretty sure has gone on tirades about the importance of semi-colons.

Mikeysax commented:
oh, even better! hah

Commit:
Fixed linting issue

Fixed linting, added command to package.json for linting. Components have prop validation.

Mikeysax commented:
This should be good now.

Mikeysax commented:
Where are the thumbs at? @kenmazaika @lasley lol

kenmazaika commented:
I like the changes.  Linting fixed, fixing linting issues and I like that you added proptype validation.  :+1: from me.

lasley commented:
Haven't had time to review yet- at a conference

lasley commented:
Whoops, fat finger cell phone git status! I'll have some time to review Saturday- at a conference this week

kenmazaika commented:
> Whoops, fat fiver cell phone git status!

Best. GitHub.  Comment.  Ever.

lasley commented:
Lol especially after accidentally closing the PR with said "fat fiver" abilities 😛

lasley commented on src/js/containers/Home.js:
```diff
@@ -0,0 +1,42 @@
+    const posts = this.props.posts.map((post, index) => <Post key={index} index={index} post={post} />);
+
+    return (
+      <div>
```
FYI you might notice inconsistent styling by nesting Bootstrap columns w/o using the row div container. It works now, but will not scale when more rows are introduced

Mikeysax replied:
Is that what you mean? the latest push.

lasley replied:
Yes. Although now that I am looking, does this actually work? The HTML attribute to control CSS classes is `class` such as `<div class="row">`.

Mikeysax replied:
In React it's called JSX which looks like HTML but renders to JavaScript. The className is used because class is a keyword. It should work I think.

lasley replied:
Sweet I had a feeling that was the case, just wanted to make sure

Commit:
Fixed bootstrap row forrendering children post components

lasley commented:
👍

lasley commented:
Good job on this one @Mikeysax - second time is the charm 😉

Mikeysax commented:
Hah thanks, I blame you! 😎
